[BLOG] PWA install on Sonoma (#4645)
pwa-builder/PWABuilder: apps/blog/src/posts/macos-pwa-install/macos-pwa-install.md

---
layout: post
title: Web Apps on macOS Sonoma got a proper install experience

> excerpt: This post guides us how we detected Web Apps availability on macOS platform.

Reviewer: ```suggestion
excerpt: How we detected Web Apps availability on macOS platform.
```

description: How we enabled Sonoma's Web Apps support for PWA Install component.
date: 2024-02-01
updatedDate: 2024-02-01
trending: true
featured: true
image: posts/macos-pwa-install/pic-1-wide.jpg
isPost: true
backUrl: '/'
author:
  name: Gleb Khmyznikov
  twitter: khmyznikov
  title: Software Engineer
  image: /author_images/gleb_image.jpg
  tagline: Eastern Europe based Software Engineer who loves PC hardware, gaming handhelds, classic cars and web technologies.
tags:
  - post
  - macOS

>   - WebApps

Reviewer 2: Web Apps

---

<figure>
  <video preload="none" controls poster="/posts/macos-pwa-install/pic-1-wide.jpg">
    <source src="/posts/macos-pwa-install/video.webm" type="video/webm">
  </video>

>   <figcaption>Installing PWA on macOS Safari</figcaption>

Reviewer 3: PWA's

</figure>

### Backstory

Traditionally for Apple, the company has been slow to adopt new web technologies. However, with the release of macOS Sonoma, Apple has made a big step forward in supporting Web Apps in general and Progressive Web Applications specifically. This is great news for developers and users alike, as it means that PWAs will be more accessible on Mac than ever before.

> But, also traditionally for Apple, there is no built-in browsers experiences for prompting PWA installation on Safari. This is where the [pwa-install](https://github.com/khmyznikov/pwa-install) component comes in. It's a simple, lightweight, framework agnostic web-component that provides a native-like installation experience for PWAs on iOS/iPadOS and now on macOS Sonoma. It's easy to use and works with any PWA or simple Web App, so you can get started right away.

Reviewer 2: I don't think the start of the first sentence flows naturally. I get you are trying to emphasize that its also traditional for them to not have built in browser experiences for prompting install, but there has to be a better way to start that sentence

> But, фccording to the unfortunate tradition of Apple, there is no built-in browsers experiences for prompting PWA installation on Safari. This is where the [pwa-install](https://github.com/khmyznikov/pwa-install) component comes in. It's a simple, lightweight, framework agnostic web-component that provides a native-like installation experience for PWAs on iOS/iPadOS and now on macOS Sonoma. It's easy to use and works with any PWA or simple Web App, so you can get started right away.

Reviewer 3: According needs an `a`

> But, фccording to the unfortunate tradition of Apple, there is no built-in browsers experiences for prompting PWA installation on Safari. This is where the [pwa-install](https://github.com/khmyznikov/pwa-install) component comes in. It's a simple, lightweight, framework agnostic web-component that provides a native-like installation experience for PWAs on iOS/iPadOS and now on macOS Sonoma. It's easy to use and works with any PWA or simple Web App, so you can get started right away.

Reviewer 3: If the web app does not have a web manifest, does the component work?

Author: Yes, but you need to manually pass the name icons etc in this case.

> But, according to the unfortunate tradition of Apple, there is no built-in browsers experiences for prompting PWA installation on Safari. This is where the [pwa-install](https://github.com/khmyznikov/pwa-install) component comes in. It's a simple, lightweight, framework agnostic web-component that provides a native-like installation experience for PWAs on iOS/iPadOS and now on macOS Sonoma. It's easy to use and works with any PWA or simple Web App, so you can get started right away.

Reviewer: I don't think we make it clear enough that Apple is doing a shitty job LOL

![PWA Install Instructions](/posts/macos-pwa-install/pic-3.jpg)

> Why easy installation matters? Progressive Web Applications offer a lot of flexibility and user engagement. One of the main selling points is their app-like presence on your device. However, not every browser has made it easy to install them. This update is set to make the installation of PWAs feel more native, especially for users on macOS Safari.

Reviewer 2: ```suggestion
Why does easy installation matter? Progressive Web Applications offer a lot of flexibility and user engagement. One of the main selling points is their app-like presence on your device. However, not every browser has made it easy to install them. This update is set to make the installation of PWAs feel more native, especially for users on macOS Safari.
```

### Implementation

> But how it was archived? Safari on macOS Sonoma supports the Web App Manifest, which is a JSON file that describes the Web App metadata. This includes the app's name, description, icons, and more. The pwa-install component uses this information to create a native-like installation experience for PWAs on iOS/iPadOS and macOS Sonoma.

Reviewer 2: ```suggestion
But how is it archived? Safari on macOS Sonoma supports the Web App Manifest, which is a JSON file that describes the Web App metadata. This includes the app's name, description, icons, and more. The pwa-install component uses this information to create a native-like installation experience for PWAs on iOS/iPadOS and macOS Sonoma.
```

Also, how is what archived? I feel like asking that question is as if the reader has this question already, but as the reader I am questioning what it is that is being archived.

Author: Any suggestions how to rewrite?

Reviewer 2: I think leaving the question format is fine. I left a suggestion for how to change the wording of the question

> But here's a catch: Safari on macOS don't provide any API to detect Web Apps availability. So we had to use a workaround to detect Web Apps availability on macOS platform.

Reviewer 2: ```suggestion
But here's a catch: Safari on macOS doesn't provide any API to detect Web Apps availability. So we had to use a workaround to detect Web Apps availability on macOS platform.
```

Reviewer 2: I am confused here about what "availability" means. Is it availability for install?

Author: Yeah it's about installability

> [Here's how it works](https://github.com/khmyznikov/pwa-install/blob/cf73d0c382fd87aa6b5a5cc40f0474150efe3487/src/utils.ts#L24):

Reviewer 2: Do we need to link to the code base if we are going to put the code directly below anyway?

Author: Why not? Maybe someone will be in more comfort with reading it on GH instead of article's code view

```js
static isAppleDesktop(): boolean {
    // check if it's a mac
    const userAgent = navigator.userAgent.toLowerCase();
    if (navigator.maxTouchPoints || !userAgent.match(/macintosh/))
        return false;
    // check safari version >= 17
    const version = /version\/(\d{2})\./.exec(userAgent)
    if (!version || !version[1] || !(parseInt(version[1]) >= 17))
        return false;
    // hacky way to detect Sonoma
    const audioCheck = document.createElement('audio').canPlayType('audio/wav; codecs="1"') ? true : false;
    const webGLCheck = new OffscreenCanvas(1, 1).getContext('webgl') ? true : false;

    return audioCheck && webGLCheck;
}
```

> Combined the user agent and specific for Safari on Sonoma feature detections, we can detect Web Apps availability on macOS platform. We had to run previous and current macOS version with latest Safari 17 side by side and with help of [tool like this](https://browserleaks.com/features) we were able to detect the difference we can use.

Reviewer 2: ```suggestion
When we combine the user agent and Sonoma specific Safari feature detections, we can detect Web Apps availability on macOS platform. We had to run previous and current macOS version with latest Safari 17 side by side and with help of [tool like this](https://browserleaks.com/features) we were able to detect the difference we can use.
```

> When we combine the user agent and Sonoma specific Safari feature detections, we can detect Web Apps availability on macOS platform. We had to run previous and current macOS version with latest Safari 17 side by side and with help of [tool like this](https://browserleaks.com/features) we were able to detect the difference we can use.

Reviewer 3: Needs a comma after the link

### Installation

> This isn't just about Safari on MacOS. The component keeps the user experience consistent no matter the device.

Reviewer 2: Sometimes you say "MacOS" and other times you say "macOS" we should figure out which one is correct for Apple and use that all the way through.

| &nbsp;&nbsp;iOS&nbsp;&nbsp; | &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | Android |
| --- | --- | --- |
|![iOS install prompt](/posts/macos-pwa-install/pic-4.jpg)|![iOS install instructions](/posts/macos-pwa-install/pic-5.jpg)|![Android Gallery](/posts/macos-pwa-install/pic-6.jpg)|

> ![Chrome Prompt](/posts/macos-pwa-install/pic-7.jpg)

Reviewer 2: Should we show the prompt on Edge instead of on Chrome?

Author: Hehe nice catch

Adding this component to your project is simple. The [readme](https://github.com/khmyznikov/pwa-install?tab=readme-ov-file#install) includes a one-line npm install command, import instructions, and basic HTML for adding the component to a app. There's even a live [demo](https://khmyznikov.com/pwa-install/) to show you exactly how it should look and work.

> React polyfill is also available for those who use React. Another modern framework users can use the component as a web component.

Reviewer 2: > Another modern framework users can use the component as a web component.

That part didn't read well to me

Author: How is now?

> What's Coming Soon for PWA Installation? Samsung Internet and Firefox Mobile are the next browsers to get the pwa-install component. This will make it even easier for users on Android devices to install PWAs from their favorite browsers.

Reviewer 2: ```suggestion
What's coming soon for PWA installation? Samsung Internet and Firefox Mobile are the next browsers to get the pwa-install component. This will make it even easier for users on Android devices to install PWAs from their favorite browsers.
```

### Conclusion

> The pwa-install component is improving the install process for PWAs, particularly with the latest Safari update for MacOS Sonoma. This update promises to make the experience feel more native for Apple users. Whether you're aiming to reach MacOS, iOS, or wider audiences, the component is essential for seamless PWA installations.

Reviewer 2: Normally, I wouldn't use such matter-of-fact statements surrounding subjective statements. 
```suggestion
The pwa-install component aims to improve the install process for PWAs, with the most recent update making the experience feel more native for Safari users on the new MacOS Sonoma. Whether you're aiming to reach MacOS, iOS, or wider audiences, the component is essential for seamless PWA installations.
```

> Jump into the [demo](https://khmyznikov.com/pwa-install/), see the new features for yourself, and if you like the progress, feel free to [contribute](https://github.com/khmyznikov/pwa-install) features or translations through pull requests.

Reviewer 2: ```suggestion
Jump into the [demo](https://khmyznikov.com/pwa-install/), see the new features for yourself,. If you like the progress, feel free to [contribute](https://github.com/khmyznikov/pwa-install) features or translations through pull requests.
```
